File: UDF Dönüştürücü/Core/Tools/UDFEditModel.swift
import Foundation
import UIKit

// MARK: - Düzenlenebilir UDF belge modeli

struct UDFEditDocument: Equatable {
    var headers: [UDFEditHeaderFooter] = []
    var blocks: [UDFEditBlock]
    var footers: [UDFEditHeaderFooter] = []
}

struct UDFEditHeaderFooter: Equatable, Identifiable {
    let id = UUID()
    var type: String = "default"
    var paragraphs: [UDFEditParagraph] = [UDFEditParagraph(runs: [UDFEditRun(text: "")])]
}

enum UDFEditBlock: Equatable, Identifiable {
    case paragraph(UDFEditParagraph)
    case table(UDFEditTable)

    var id: String {
        switch self {
        case .paragraph(let p): return "p-\(p.id.uuidString)"
        case .table(let t): return "t-\(t.id.uuidString)"
        }
    }
}

struct UDFEditParagraph: Equatable, Identifiable {
    let id = UUID()
    var alignment: Int = 3
    var spaceAbove: CGFloat = 1
    var spaceBelow: CGFloat = 1
    var leftIndent: CGFloat = 0
    var rightIndent: CGFloat = 0
    var firstLineIndent: CGFloat = 0
    var hangingIndent: CGFloat = 0
    var lineSpacing: CGFloat = 0
    var tabStops: [CGFloat] = []
    var runs: [UDFEditRun] = []
}

struct UDFEditRun: Equatable {
    enum Kind: Equatable {
        case content
        case field(name: String)
        case space
    }

    var kind: Kind = .content
    var text: String
    var isBold: Bool = false
    var isItalic: Bool = false
    var isUnderline: Bool = false
    var fontSize: CGFloat = 12
    var fontFamily: String = "Times New Roman"
    /// Java/UYAP signed ARGB (siyah = -16777216)
    var foregroundARGB: Int?
    var backgroundARGB: Int?

    var isField: Bool {
        if case .field = kind { return true }
        return false
    }

    var fieldName: String? {
        if case .field(let name) = kind { return name }
        return nil
    }
}

struct UDFEditTable: Equatable, Identifiable {
    let id = UUID()
    var columnCount: Int = 2
    var columnSpans: [Int] = []
    var border: String = "borderCell"
    var rows: [UDFEditTableRow] = []
}

struct UDFEditTableRow: Equatable, Identifiable {
    let id = UUID()
    var rowType: String = "dataRow"
    var cells: [UDFEditTableCell] = []
}

struct UDFEditTableCell: Equatable, Identifiable {
    let id = UUID()
    var colspan: Int = 1
    var rowspan: Int = 1
    var fillColorARGB: Int?
    var paragraphs: [UDFEditParagraph] = [UDFEditParagraph(runs: [UDFEditRun(text: "")])]

    /// Birleştirilmiş hücrelerde ilk paragrafı döndürür.
    func primaryParagraphValue() -> UDFEditParagraph {
        if paragraphs.isEmpty {
            return UDFEditParagraph(runs: [UDFEditRun(text: "")])
        }
        return paragraphs[0]
    }

    mutating func setPrimaryParagraph(_ paragraph: UDFEditParagraph) {
        if paragraphs.isEmpty {
            paragraphs = [paragraph]
        } else {
            paragraphs[0] = paragraph
        }
    }
}

// MARK: - Renk yardımcıları

enum UDFColorCodec {
    static func uiColor(fromJavaARGB value: Int) -> UIColor {
        let u = UInt32(bitPattern: Int32(value))
        let a = CGFloat((u >> 24) & 0xFF) / 255
        let r = CGFloat((u >> 16) & 0xFF) / 255
        let g = CGFloat((u >> 8) & 0xFF) / 255
        let b = CGFloat(u & 0xFF) / 255
        return UIColor(red: r, green: g, blue: b, alpha: max(a, 0.01))
    }

    static func javaARGB(from color: UIColor) -> Int {
        var r: CGFloat = 0, g: CGFloat = 0, b: CGFloat = 0, a: CGFloat = 1
        color.getRed(&r, green: &g, blue: &b, alpha: &a)
        return (Int(a * 255) << 24) | (Int(r * 255) << 16) | (Int(g * 255) << 8) | Int(b * 255)
    }

    static func parse(_ raw: String?) -> Int? {
        guard let raw, !raw.isEmpty else { return nil }
        if let intVal = Int(raw.trimmingCharacters(in: .whitespaces)) {
            return intVal
        }
        let parts = raw.split(separator: ",").compactMap { Int($0.trimmingCharacters(in: .whitespaces)) }
        guard parts.count >= 3 else { return nil }
        return (255 << 24) | (parts[0] << 16) | (parts[1] << 8) | parts[2]
    }

    static func format(_ argb: Int) -> String {
        String(argb)
    }
}

extension NSAttributedString.Key {
    static let udfFieldName = NSAttributedString.Key("udfFieldName")
}

File: android/app/src/main/java/com/velikececi/udfdonusturucu/core/tools/editor/UdfEditModel.kt
package com.velikececi.udfdonusturucu.core.tools.editor

import java.util.UUID

/** UDFEditModel.swift'in Kotlin karşılığı — UDF Düzenleyici'nin düzenlenebilir belge modeli. */
data class UdfEditDocument(
    val headers: List<UdfEditHeaderFooter> = emptyList(),
    val blocks: List<UdfEditBlock>,
    val footers: List<UdfEditHeaderFooter> = emptyList(),
)

data class UdfEditHeaderFooter(
    val id: String = UUID.randomUUID().toString(),
    val type: String = "default",
    val paragraphs: List<UdfEditParagraph> = listOf(UdfEditParagraph(runs = listOf(UdfEditRun(text = "")))),
)

sealed class UdfEditBlock {
    abstract val id: String

    data class ParagraphBlock(val paragraph: UdfEditParagraph) : UdfEditBlock() {
        override val id: String get() = "p-${paragraph.id}"
    }

    data class TableBlock(val table: UdfEditTable) : UdfEditBlock() {
        override val id: String get() = "t-${table.id}"
    }
}

data class UdfEditParagraph(
    val id: String = UUID.randomUUID().toString(),
    val alignment: Int = 3, // 0=sol, 1=orta, 2=sağ, 3=iki yana yasla
    val spaceAbove: Float = 1f,
    val spaceBelow: Float = 1f,
    val leftIndent: Float = 0f,
    val rightIndent: Float = 0f,
    val firstLineIndent: Float = 0f,
    val hangingIndent: Float = 0f,
    val lineSpacing: Float = 0f,
    val tabStops: List<Float> = emptyList(),
    val runs: List<UdfEditRun> = emptyList(),
)

sealed class UdfRunKind {
    data object Content : UdfRunKind()
    data class Field(val name: String) : UdfRunKind()
    data object Space : UdfRunKind()
}

data class UdfEditRun(
    val kind: UdfRunKind = UdfRunKind.Content,
    val text: String,
    val isBold: Boolean = false,
    val isItalic: Boolean = false,
    val isUnderline: Boolean = false,
    val fontSize: Float = 12f,
    val fontFamily: String = "Times New Roman",
    /** Java/UYAP işaretli (signed) ARGB — siyah = -16777216. Android `Color` int'iyle bit-bit aynı. */
    val foregroundARGB: Int? = null,
    val backgroundARGB: Int? = null,
) {
    val isField: Boolean get() = kind is UdfRunKind.Field
    val fieldName: String? get() = (kind as? UdfRunKind.Field)?.name
}

data class UdfEditTable(
    val id: String = UUID.randomUUID().toString(),
    val columnCount: Int = 2,
    val columnSpans: List<Int> = emptyList(),
    val border: String = "borderCell",
    val rows: List<UdfEditTableRow> = emptyList(),
)

data class UdfEditTableRow(
    val id: String = UUID.randomUUID().toString(),
    val rowType: String = "dataRow",
    val cells: List<UdfEditTableCell> = emptyList(),
)

data class UdfEditTableCell(
    val id: String = UUID.randomUUID().toString(),
    val colspan: Int = 1,
    val rowspan: Int = 1,
    val fillColorARGB: Int? = null,
    val paragraphs: List<UdfEditParagraph> = listOf(UdfEditParagraph(runs = listOf(UdfEditRun(text = "")))),
) {
    /** Birleştirilmiş hücrelerde ilk paragrafı döndürür. */
    fun primaryParagraphValue(): UdfEditParagraph =
        paragraphs.firstOrNull() ?: UdfEditParagraph(runs = listOf(UdfEditRun(text = "")))

    fun withPrimaryParagraph(paragraph: UdfEditParagraph): UdfEditTableCell {
        val newParagraphs = if (paragraphs.isEmpty()) {
            listOf(paragraph)
        } else {
            paragraphs.toMutableList().also { it[0] = paragraph }
        }
        return copy(paragraphs = newParagraphs)
    }
}

/**
 * UDFColorCodec.swift'in Kotlin karşılığı. iOS tarafı Java işaretli ARGB'yi `UIColor`'a
 * dönüştürmek zorundaydı; Android `Color` int'i zaten aynı bit düzenine (0xAARRGGBB, işaretli
 * Int) sahip olduğundan bu dönüşüm gerekmez — model değeri render katmanında doğrudan
 * `android.graphics.Color`/span rengi olarak kullanılabilir (bkz. `UdfSpanCodec`).
 */
object UdfColorCodec {
    fun parse(raw: String?): Int? {
        if (raw.isNullOrEmpty()) return null
        raw.trim().toIntOrNull()?.let { return it }
        val parts = raw.split(",").mapNotNull { it.trim().toIntOrNull() }
        if (parts.size < 3) return null
        return (255 shl 24) or (parts[0] shl 16) or (parts[1] shl 8) or parts[2]
    }

    fun format(argb: Int): String = argb.toString()
}
